zcash_primitives/src/transaction/builder.rs

```diff
@@ -716,6 +731,10 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
                     spend_auth_sig: None,
                 });
 
+                // Update progress and send a notification on the channel
+                progress += 1;
+                tx.as_ref().map(|tx| tx.send(progress));
```

Comment:
As a second thought, what if instead of passing a `Sender` to the `build` function, you were to simply pass a `FnMut` callback that could be invoked at  each step? That way the choice of channel is up to the caller, and it's possible to use in contexts that don't support asynchrony for collecting other metrics.

Reply:
I didn't do this because I thought it would break the builder API. I think Nighthawk and Unstoppable are also using this API, and it would break for them. 

I originally had a `FnMut` passed in, but I thought it would be a bad idea, because the caller might do arbitrary computations in the callback, possibly even blocking it. This would slow down the builder. I thought using a `Sender` is safer against bugs in the callback code.